### ISSUE TITLE
Not possible to trace deleted features by the "FeatureRemoved" Event

### ISSUE THREAD

**Issue by melofranco** [DotSpatial/DotSpatial#1205]
*IF BUG, INCLUDE THIS PART:*

### Steps to reproduce

1. Create an instance of a FeatureSet ralated to a FeatureLayer already added to a map
2. Create a method to capture FeatureRemoved Event and display the Fid of the Feature
3. Select some features on the the layer and delete them

DotSpatial version: 2.0

### Expected behaviour
        Displaying the Fid of the deleted features

Tell us what should happen

### Actual behaviour
        An exception occurs telling that the instance of the features doesn't exist

Tell us what happens instead
Can you also include a screen shot?

*IF IT IS A NEW FEATURE REQUEST, INCLUDE THIS PART:*

### Feature description
        Reading the source code of the EventHandler, i notice that the feature is deleted before issuing the Event. Is it possible to invert the process ?

Write a description of the feature. How should it work? How should it look?
Include some graphics if this could help!

**Comment by VectorZita** [comment]
It's not like your suggestion is wrong, but there is a specific logic for that, as far as I understand. The **FeatureRemoved** event is raised when a feature has been removed, meaning (literally) you cannot "undo" the process while at it (as would be possible if using a **FeatureRemoving** event with suitable EventArgs to cancel).

This, however, is irrelevant. The problem with your request is that if you would display the FID of the feature that was removed, you would be displaying an FID that is useless for any specific purpose thereafter, as it is now referring to a **different feature**. FID values make sense as long as a feature is **within** a **FeatureSet**, not for "stray" ones. When you remove a feature, another feature potentially (if the removed feature was not the last one) takes its FID value, so, whichever the FID value, it is invalidated as soon as you remove the feature, and you can not do anything meaningful with it thereafter.

In short, FID values and features go hand in hand only within the limits of a feature set. Upon removal, the previous FID of the feature makes no sense anymore to **that specific feature**. Besides, if you wish to display the FID of the feature, you can definitely do so prior to removal, as the removal action has to be initiated somewhere, and this is something you will have to handle in your code. As such, you can intercept your own code at the points of your "delegated (requested) removal" and display the FID there, if you need to, i.e. wrap the removal process in an event of your own.

**Comment by melofranco** [comment]
So, is it possible to fire the event before deleting feature ?

**Comment by VectorZita** [comment]
As it is now, only if you change the source code of the core file (FeatureSet.cs) and rebuilding. However, this would be easier and cleaner to do in your own code. You obviously have some calls to FeatureSet.Remove somewhere, right? Whenever you call Remove, you can create a different function that you can call, for example, RemoveFeature(FeatureSet fs, IFeature f), which wraps the procedure and fires some custom event you have created, which you can then handle wherever you wish in your code. 

It depends a lot on what you want to do and what you already have in terms of code in specific, of course, but it is practically doable without having to alter the DotSpatial codebase and recompile at all.
